### PR TITLE
Allow building Exiv2 0.27 with C++17 compiler

### PR DESCRIPTION
**Pull request by maksim-petukhou**
The only thing that is preventing this from happening is using `std::auto_ptr`. This PR introduces two compile definitions `EXV_NO_AUTO_PTR` (I looked at `BOOST_NO_AUTO_PTR` for inspiration) and `EXV_NO_AUTO_PTR_MOVE` which are defined only when compiling library with C++17 compiler flag. As far as I can tell there are no existing users of this library which are affected by this change: even using C++11 and C++14 compiler flag user will have the same behavior as before this PR, `std::auto_ptr` will be used. Looks not really nice, I know, but I think it is OK to have this on maintenance branch. I tested this changes locally with MSVC 192, gcc 10.4 and clang from android sdk r25c.

### CODE REVIEW

**Comment by postscript-dev** [comment]
@maksim-petukhov 
The `0.27-maintenance` branch is fixed using C++98 and, as far as I know, there are no plans to change that. However recently, Exiv2 released version [0.28.0](https://exiv2.org/download.html) which is compatible with C++20.

Hopefully the new release provides the solution that you need.

**Comment by maksim-petukhov** [comment]
@postscript-dev As I said in the first post, this doesn't change anything for existing users of library. It is still good old C++98 when they do not use C++17 compiler mode. And there are no users who used C++17 standard with 0.27.x because it wasn't possible prior that PR. I tried to use Exiv2 0.28, but unfortunately in my project I can't open files with non-ASCII paths using it (see [my comment](https://github.com/Exiv2/exiv2/issues/2637#issuecomment-1679383461))

**Comment by neheb** [comment]
this is an API change.

I've seen libcxx fail with compilation as _LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR is needed.

**Comment by maksim-petukhov** [comment]
@neheb I'm sorry, but how does it change the API?

**Comment by neheb** [comment]
You’re redefining AutoPtr as unique_ptr.

**Comment by maksim-petukhov** [comment]
> You’re redefining AutoPtr as unique_ptr.

Yes, but I'm doing it only for C++17 compiler mode so no users are affected by this API change. This library cannot be used with C++17 now, so there is no API at all.

**Comment by postscript-dev** [comment]
@maksim-petukhov 
Whether this PR will be merged or not, I don't know. However, there are some things to bare in mind.

1. As far as I understand it, the `0.27-maintenance` branch is not under any active development and will only receive limited security fixes at best. 

2. As the project's resources are limited, our focus is on Exiv2 `0.28.x`. I think that we would usually encourage users to upgrade rather than work on an old version. Although upgrading is not an option for you.

3. Exiv2 `0.27.x` has been removed from the project's CI meaning we are not fuzzing, checking for any build issues or running the Exiv2 testing.

To summarise, although your PR might provide a solution it will cause several maintenance issues moving forward.

**Comment by neheb** [comment]
google says for MSVC, _HAS_AUTO_PTR_ETC=1 needs to be defined. These two changes need to be done with CMakeLists.txt , not by changing headers and whatnot.

**Comment by kmilos** [comment]
> These two changes need to be done with CMakeLists.txt

Not even, you can define on the command line when calling CMake...

**Comment by maksim-petukhov** [comment]
@neheb @kmilos These defines are required not only to build exiv2. You have to define them for you project too because exiv2 uses auto_ptr in its interface. This can be OK for some "Hello World"-like program but not really appropriate for production code. There are internal guidelines, there are tendencies for conformance with C++ standard and these compiler defines are not compatible with these things. We can see how other people do this:
- freeimage patched both [in conan](https://github.com/conan-io/conan-center-index/blob/237d3a8e1607e48a7114ac2d3a69032797e78853/recipes/freeimage/all/patches/014_no_auto_ptr.patch#L7) and [vcpkg](https://github.com/microsoft/vcpkg/pull/26007/files#diff-4531f58b8e1c9e666e9c753ddc4ed755000dc96792afde73a1d50531a71a7356) to use unique_ptr instead of auto_ptr
- podofo patched [in vcpkg](https://github.com/microsoft/vcpkg/pull/5182/files) to use unique_ptr instead of auto_ptr
- libxmlpp patched [in vcpkg](https://github.com/microsoft/vcpkg/pull/3317/files#diff-3dcb16cfbcea972b59c0fa54d2d83d424ff121a7a77783cc1e25682622634bca) to use unique_ptr instead of auto_ptr
- coin-lemon patched [in conan](https://github.com/conan-io/conan-center-index/blob/master/recipes/coin-lemon/all/patches/0001-remove-register-keyword.patch) to remove `register` keyword
- boost have similar `BOOST_NO_AUTO_PTR` to use unique_ptr instead of auto_ptr ([see boost graph code](https://github.com/boostorg/graph/blob/942829ee68304438df773264484761244972ec08/include/boost/graph/detail/adjacency_list.hpp#L290C5-L290C5) for example)

Some of these libraries are not maintained AFAIK so you **have to** patch them. Having these necessary changes within the library itself is a more streamlined approach. This is the motivation behind my PR.
